src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs

```diff
@@ -78,6 +78,91 @@ public static string CreateClientId(string prefix)
+        {
+            if (!RemoteSupportedLanguages.IsSupported(project.Language))
+            {
+                return SpecializedTasks.Default<RemoteHostClient?>();
+            }
+
+            return TryGetClientAsync(project.Solution.Workspace, cancellationToken);
+        }
+
+        public static Task<RemoteHostClient?> TryGetClientAsync(Workspace workspace, CancellationToken cancellationToken)
+        {
+            var service = workspace.Services.GetService<IRemoteHostClientService>();
+            if (service == null)
+            {
+                return SpecializedTasks.Default<RemoteHostClient?>();
+            }
+
+            return service.TryGetRemoteHostClientAsync(cancellationToken);
+        }
+
+        /// <summary>
+        /// Creates <see cref="SessionWithSolution"/> for the <paramref name="serviceName"/> if possible, otherwise returns <see langword="null"/>.
+        /// </summary>
+        public async Task<SessionWithSolution?> TryCreateSessionAsync(string serviceName, Solution solution, object? callbackTarget, CancellationToken cancellationToken)
+        {
+            var connection = await TryCreateConnectionAsync(serviceName, callbackTarget, cancellationToken).ConfigureAwait(false);
+            if (connection == null)
+            {
+                return null;
+            }
+
+            return await SessionWithSolution.CreateAsync(connection, solution, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Creates <see cref="KeepAliveSession"/> for the <paramref name="serviceName"/>, otherwise returns <see langword="null"/>.
+        /// </summary>
+        public async Task<KeepAliveSession?> TryCreateKeepAliveSessionAsync(string serviceName, object? callbackTarget, CancellationToken cancellationToken)
+        {
+            var connection = await TryCreateConnectionAsync(serviceName, callbackTarget, cancellationToken).ConfigureAwait(false);
+            if (connection == null)
+            {
+                return null;
+            }
+
+            return new KeepAliveSession(this, connection, serviceName, callbackTarget);
+        }
+
+        public async Task<bool> TryRunRemoteAsync(string serviceName, string targetName, Solution solution,
+            IReadOnlyList<object> arguments, object? callbackTarget, CancellationToken cancellationToken)
+        {
+            using var session = await TryCreateSessionAsync(serviceName, solution, callbackTarget, cancellationToken).ConfigureAwait(false);
+            if (session == null)
+            {
+                return false;
+            }
+
+            await session.InvokeAsync(targetName, arguments, cancellationToken).ConfigureAwait(false);
+            return true;
+        }
+
+        public async Task<bool> TryRunRemoteAsync(string serviceName, string targetName, IReadOnlyList<object> arguments, object? callbackTarget, CancellationToken cancellationToken)
```

Comment:
the difference between this method and the one above is super subtle.  it looks like one is just missing a Solution param.  Seems like it would be easy for someone to accidentcally call the method with less args (which seems like the *wrong* one most of the time).  

Can we rename these to make it clear that they're very different and one shoudl be extremey careful about which is called?

Reply:
I agree, but can't think of a good name for the method. How about making the solution nullable (but required) so the call-sites that do not pass solution would need to explicitly specify null?

Reply:
Also there is https://github.com/dotnet/roslyn/issues/24836

Reply:
> I agree, but can't think of a good name for the method. How about making the solution nullable (but required) so the call-sites that do not pass solution would need to explicitly specify null?

I like that idea.  We should comment those points (where null is passed) as well to indicate why we're getting the appropriate semantics we want there.